internal/healthchecks/generic/generic.go

```diff
@@ -0,0 +1,198 @@
+package generic
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AreObjectsHealthy checks if the given resources are healthy.
+// It returns true if all resources are healthy with nil error.
+// It returns false if any of the resources are not healthy, the error contains the GVK + resourceName and the error message of
+// each unhealthy resource.
+//
+// The current list of supported resources is:
+// - Deployments
+// - StatefulSets
+// - DaemonSets
+// - ReplicaSets
+// - Pods
+// - APIServices
+// - CustomResourceDefinitions
+//
+// If the resource is not supported, it is assumed to be healthy.
+func AreObjectsHealthy(ctx context.Context, client client.Client, objects []client.Object) (bool, error) {
+	gvkErrorsMap := make(map[string]string)
+
+	// Create a new scheme to be able to convert the objects to their correct version.
+	s := runtime.NewScheme()
+	apiregistrationv1.AddToScheme(s)
+	apiextensionsv1.AddToScheme(s)
+	appsv1.AddToScheme(s)
+	corev1.AddToScheme(s)
+
+	for _, object := range objects {
+		// Convert the object to its proper type, deepCopy it to avoid mutating the original object.
+		obj, err := runtime.ObjectConvertor(s).ConvertToVersion(object.DeepCopyObject(), object.GetObjectKind().GroupVersionKind().GroupVersion())
+		if err != nil {
+			gvkErrorsMap[toErrKey(object)] = err.Error()
+			continue
+		}
+		switch obj.(type) {
+		case *appsv1.Deployment:
+			deployment := obj.(*appsv1.Deployment)
+
+			if err := client.Get(ctx, types.NamespacedName{
+				Name:      deployment.GetName(),
+				Namespace: deployment.GetNamespace(),
+			}, deployment); err != nil {
+				gvkErrorsMap[toErrKey(deployment)] = err.Error()
+				continue
+			}
+
+			if deployment.Status.Conditions == nil {
+				gvkErrorsMap[toErrKey(deployment)] = "Deployment has no conditions"
+				continue
+			}
+
+			for _, condition := range deployment.Status.Conditions {
```

Comment:
Global comment: I would modify these condition checks to explicitly look for the expected condition and its expected status. For example, in this check, if there is no DeploymentAvailable condition, it is still considered healthy.

Reply:
done